Mirobot.cpp

```diff
@@ -116,6 +116,107 @@ void Mirobot::right(int angle){
+
+  // work out whether we're doing clockwise or anticlockwise arc
+  clockwise = radius > 0 ? false : true;
+  radius = abs(radius);
+
+  /* Showing the working, then the shortcuts for the code!
+
+     When drawing a circle with the pen, one turtle wheel will be walking
+     a bigger (outer) circle and the other a smaller (inner) circle.
+
+     Calculate radius of inner/outer circle using known wheel distance:
+       inner_circle_radius = radius - (WHEEL_DISTANCE / 2.0)
+       outer_circle_radius = radius + (WHEEL_DISTANCE / 2.0)
+
+     Now calculate circumference of inner/outer circle: 2r * pi
+       inner_circle_circumference = (2 * inner_circle_radius * 3.1416)
+       outer_circle_circumference = (2 * outer_circle_radius * 3.1416)
+
+     Now calculate distance required for angle given
+       inner_circle_distance = (inner_circle_circumference / 360) * angle
+       outer_circle_distance = (outer_circle_circumference / 360) * angle
+
+     We can shortcut (2 * radius * pi) / 360, because
+     2 * pi / 360 = pi / 180 = 0.017453333
+     This is known as converting degrees to radians: radius / 0.017453333
+
+     An arc of a circle with the same length as the radius of that circle
+     corresponds to an angle of 1 radian. A full circle corresponds to an
+     angle of 2π radians.
+     Ref: https://en.wikipedia.org/wiki/Radian
+
+     So we can say distance = (radius / 0.017453333) * angle
+     and bypass the circumference bit in the middle of our calculation.
+ */
+
+  // Calculate radius of inner/outer circle using known wheel distance
+  inner_circle_radius = radius - (WHEEL_DISTANCE / 2.0);
+  outer_circle_radius = radius + (WHEEL_DISTANCE / 2.0);
+
+  // Now calculate how far one radian is for ecah circle
+  inner_circle_radian = inner_circle_radius / 0.017453333;
+  outer_circle_radian = outer_circle_radius / 0.017453333;
+
+  // If circle radius is smaller than distance between pen and wheel,
+  // we need to actually run the inner wheel backwards!
+  if (inner_circle_radian < 0) {
+    smallcircle = true;
+    inner_circle_radian = abs(inner_circle_radian);
+  }
+
+  /* Now comes the real tricky bit. It's not the maths but the movement.
+
+     The following code may change as we work out the best way of doing it.
+
+     Ideally we'd make the speed adjustable per stepper, following
+     the inner:outer distance ratio. That'll make it look good!
+
+     For our now we'll just do the arc or circle in radian bits to "fake" it.
+  */
+
+  while (angle-- > 0) {
+    // Note: motor1 on right, motor2 on left (reversed so backward moves forwards)
+    if (clockwise) {
+      motor1.turn(inner_circle_radian * STEPS_PER_MM, smallcircle ? BACKWARD : FORWARD);
+      motor2.turn(outer_circle_radian * STEPS_PER_MM, BACKWARD);
+    }
+    else {
+      motor1.turn(outer_circle_radian * STEPS_PER_MM, FORWARD);
+      motor2.turn(inner_circle_radian * STEPS_PER_MM, smallcircle ? FORWARD : BACKWARD);
+    }
+
+    wait();
```

Comment:
@arjenlentz I'm leaving a note here even though this PR is closed because it makes sense to be able to discuss somewhere.

check the definition of wait() on L289 - it's used when the code is put into blocking mode (i.e. you're programming the movements directly int he Arduino) so that it only returns when it's complete. In normal mode (i.e. the Mirobot firmware) we're non-blocking so we can do things like process serial data while the steppers are turning so this returns immediately and the distance the motors should turn is overwritten each time.

I think there are two approaches to solving this;
 - make the arc function run in a loop so that it can be called regularly from the main event loop and so once the motors have completed their movements, the next chunk is scheduled
 - make hotstepper be able to handle different speeds of rotation so there's no need to break it into small chunks

I think they're probably both about the same complexity, in which case I think it's worth just making HotStepper support variable speeds. I'll probably need to increase the interrupt rate and keep a counter so it's possible to set a rate, where 1.0 is full speed (3ms pulses) and less makes the motor turn slower. Then we could do something like calculate the desired distance each stepper should turn and then make the lesser of the two turn slower so they both finish at the same time. e.g.

```
left_wheel_distance = 100;
right_wheel_distance = 50;

if(left_wheel_distance > right_wheel_distance){
  float ratio = right_wheel_distance / left_wheel_distance;
  motor1.turn(left_wheel_distance, 1.0);
  motor2.turn(right_wheel_distance, ratio);
}else{
  float ratio = left_wheel_distance / right_wheel_distance;
  motor1.turn(left_wheel_distance, ratio);
  motor2.turn(right_wheel_distance, 1.0);
}
```

or something like that, anyway...

thoughts?